src/lib/provable/gadgets/sha2.ts

```diff
@@ -0,0 +1,621 @@
+ *
+ * @returns The updated intermediate hash values after compression.
+ */
+function compression<T extends UInt32 | UInt64>(
+  length: Length,
+  [...H]: T[],
+  W: T[]
+) {
+  let numWords = SHA2Constants.NUM_WORDS[length];
+
+  let k = SHA2Constants.K[length];
+
+  // §6.2.2.2 and §6.4.2.2:
+  // initialize working variables
+  let a = H[0];
+  let b = H[1];
+  let c = H[2];
+  let d = H[3];
+  let e = H[4];
+  let f = H[5];
+  let g = H[6];
+  let h = H[7];
+
+  // §6.2.2.3 and §6.4.2.3:
+  // main loop
+  for (let t = 0; t < numWords; t++) {
+    // T1 is unreduced and not proven to be 32 | 64 bit,
+    // we will do this later to save constraints
+    const unreducedT1 = h.value
+      .add(SigmaOne(length, e).value)
+      .add(Ch(length, e, f, g).value)
+      .add(k[t])
+      .add(W[t].value)
+      .seal();
+
+    // T2 is also unreduced
+    const unreducedT2 = SigmaZero(length, a).value.add(
+      Maj(length, a, b, c).value
+    );
+
+    h = g;
+    g = f;
+    f = e;
+    e = reduceMod(length, d.value.add(unreducedT1));
+    // mod 32 | 64 bit the unreduced field element
+    d = c;
+    c = b;
+    b = a;
+    a = reduceMod(length, unreducedT2.add(unreducedT1));
+    // mod 32 | 64 bit
+  }
+
+  // §6.2.2.4 and §6.4.2.4
+  // new intermediate hash value
+  intermediateHash(length, [a, b, c, d, e, f, g, h], H);
+
+  return H;
+}
+
+// Helper functions
+
+// Helper function to check if it is a short SHA2 (SHA2-224 or SHA2-256) or not
+function isShort(length: Length): boolean {
+  return length === 224 || length === 256;
+}
+
+// Shorthand to reduce a field element modulo 32 or 64 bits depending on T
+function reduceMod<T extends UInt32 | UInt64>(length: Length, x: Field): T {
+  if (isShort(length)) {
+    return UInt32.Unsafe.fromField(divMod32(x, 32 + 16).remainder) as T;
+  } else {
+    return UInt64.Unsafe.fromField(divMod64(x, 64 + 16).remainder) as T;
+  }
+}
+
+// Helper function to create chunks based on the size and type (UInt32 or UInt64)
+function createChunks<T extends UInt32 | UInt64>(
+  paddedMessage: UInt8[],
+  byteSize: number,
+  fromBytes: Function
+): T[] {
+  let chunks: any[] = [];
+  // bytesToWord expects little endian, so we reverse the bytes
+  for (let i = 0; i < paddedMessage.length; i += byteSize) {
+    // Chunk the data based on the specified byte size (4 bytes for UInt32, 8 bytes for UInt64)
+    chunks.push(fromBytes(paddedMessage.slice(i, i + byteSize)));
+  }
+  return chunks;
+}
+
+function intermediateHash<T extends UInt32 | UInt64>(
+  length: Length,
+  variables: T[],
+  H: T[]
+) {
+  if (isShort(length)) {
+    for (let i = 0; i < 8; i++) {
+      H[i] = (variables[i] as UInt32).addMod32(H[i] as UInt32) as T;
+    }
+  } else {
+    for (let i = 0; i < 8; i++) {
+      H[i] = (variables[i] as UInt64).addMod64(H[i] as UInt64) as T;
+    }
+  }
+}
+
+// Subroutines for SHA2
+
+function Ch<T extends UInt32 | UInt64>(length: Length, x: T, y: T, z: T): T {
```

Comment:
This works well, but if you are interested you could take a look at type guards that might make things in certain cases a bit cleaner

https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards